### PR TITLE
fixes #33

### LINKED ISSUE #33: Data race accessing variables in log.go
In https://github.com/ONSdigital/dp-search-reindex-api/pull/53/commits/14b92be2116244bb624468c82efd25b9bd2b42df you can see a commented out piece of code which was attempting to use `log.SetDestination` to restore `stdout` as the Writer used internally by our logger.

However, within `log.go` (https://github.com/ONSdigital/log.go/blob/master/log/log.go#L24) access to the module scoped `destination` and `fallbackDestination` writers is not synchronised, so

Given: logging is used by some other goroutine (e.g. dp-mongodb-in-memory)
When: `SetDestination` is called to change the log destinations
Then: A data race occurs accessing the `destination` variable

Access to module level variables needs to be synchronised (or an alternative system of instantiating a logger instance could help)

---

Example data race trace:

```
WARNING: DATA RACE
Write at 0x0000027fc8e0 by goroutine 29:
  github.com/ONSdigital/log.go/v2/log.SetDestination()
      /Users/andrewgibson/src/pkg/mod/github.com/!o!n!sdigital/log.go/v2@v2.0.9/log/log.go:310 +0x64
  github.com/ONSdigital/dp-search-reindex-api.TestComponent.func1()
      /Users/andrewgibson/src/dp-search-reindex-api/main_test.go:85 +0x25
  github.com/ONSdigital/dp-search-reindex-api.TestComponent()
      /Users/andrewgibson/src/dp-search-reindex-api/main_test.go:111 +0x75f
  testing.tRunner()
      /usr/local/Cellar/go/1.17.5/libexec/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/Cellar/go/1.17.5/libexec/src/testing/testing.go:1306 +0x47

Previous read at 0x0000027fc8e0 by goroutine 31:
  github.com/ONSdigital/log.go/v2/log.printEvent()
      /Users/andrewgibson/src/pkg/mod/github.com/!o!n!sdigital/log.go/v2@v2.0.9/log/log.go:286 +0xac
  github.com/ONSdigital/log.go/v2/log.eventWithoutOptionsCheck()
      /Users/andrewgibson/src/pkg/mod/github.com/!o!n!sdigital/log.go/v2@v2.0.9/log/log.go:207 +0x173
  github.com/ONSdigital/log.go/v2/log.Info()
      /Users/andrewgibson/src/pkg/mod/github.com/!o!n!sdigital/log.go/v2@v2.0.9/log/log.go:81 +0x72c
  github.com/ONSdigital/dp-mongodb-in-memory.stdHandler.func1()
      /Users/andrewgibson/src/pkg/mod/github.com/!o!n!sdigital/dp-mongodb-in-memory@v1.1.0/main.go:170 +0x45e

Goroutine 29 (running) created at:
  testing.(*T).Run()
      /usr/local/Cellar/go/1.17.5/libexec/src/testing/testing.go:1306 +0x726
  testing.runTests.func1()
      /usr/local/Cellar/go/1.17.5/libexec/src/testing/testing.go:1598 +0x99
  testing.tRunner()
      /usr/local/Cellar/go/1.17.5/libexec/src/testing/testing.go:1259 +0x22f
  testing.runTests()
      /usr/local/Cellar/go/1.17.5/libexec/src/testing/testing.go:1596 +0x7ca
  testing.(*M).Run()
      /usr/local/Cellar/go/1.17.5/libexec/src/testing/testing.go:1504 +0x9d1
  main.main()
      _testmain.go:43 +0x22b

Goroutine 31 (running) created at:
  github.com/ONSdigital/dp-mongodb-in-memory.stdHandler()
      /Users/andrewgibson/src/pkg/mod/github.com/!o!n!sdigital/dp-mongodb-in-memory@v1.1.0/main.go:148 +0x2f3
  github.com/ONSdigital/dp-mongodb-in-memory.Start()
      /Users/andrewgibson/src/pkg/mod/github.com/!o!n!sdigital/dp-mongodb-in-memory@v1.1.0/main.go:59 +0x4c9
  github.com/ONSdigital/dp-component-test.NewMongoFeature()
      /Users/andrewgibson/src/pkg/mod/github.com/!o!n!sdigital/dp-component-test@v0.6.3/mongo_feature.go:33 +0x84
  github.com/ONSdigital/dp-search-reindex-api.(*ComponentTest).InitializeTestSuite.func1()
      /Users/andrewgibson/src/dp-search-reindex-api/main_test.go:65 +0x47
  github.com/cucumber/godog.(*runner).scenarioConcurrent()
      /Users/andrewgibson/src/pkg/mod/github.com/cucumber/godog@v0.10.0/run.go:157 +0x2a7
  github.com/cucumber/godog.runWithOptions()
      /Users/andrewgibson/src/pkg/mod/github.com/cucumber/godog@v0.10.0/run.go:340 +0xce4
  github.com/cucumber/godog.TestSuite.Run()
      /Users/andrewgibson/src/pkg/mod/github.com/cucumber/godog@v0.10.0/run.go:445 +0x4e7
  github.com/ONSdigital/dp-search-reindex-api.TestComponent()
      /Users/andrewgibson/src/dp-search-reindex-api/main_test.go:101 +0x5a5
  testing.tRunner()
      /usr/local/Cellar/go/1.17.5/libexec/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/Cellar/go/1.17.5/libexec/src/testing/testing.go:1306 +0x47
```

### PR DESCRIPTION
**Pull request by goofballLogic**
### What

Synchronised access to destination and failureDestination writers. Fixes #33 

### How to review

Review synchronisation logic from first principles

### Who can review

Anyone comfortable with goroutines/multi-threading and syncronisation

### CODE REVIEW

File: log/log_test.go [hunk 1]
```diff
@@ -437,4 +437,53 @@ func TestLog(t *testing.T) {
...
+					SetDestination(nil, io.Discard)
+					<-letTheRaceEnd
+					So(true, ShouldEqual, true) // all we are testing for is the absence of detecting a data race
+				})
+			})
+		})
+	})
+}
+
+type WriteWillError struct {
```
Comment (nshumoogum) [inline log/log_test.go:484]: We dont need the following type, it is not used

File: log/log_test.go [hunk 2]
```diff
@@ -437,4 +437,53 @@ func TestLog(t *testing.T) {
...
+}
+
+func (w WriteWillError) Write(p []byte) (n int, err error) {
```
Comment (nshumoogum) [inline log/log_test.go:487]: The Write function is not used, can be removed

File: log/log_test.go [hunk 3]
```diff
@@ -437,4 +437,53 @@ func TestLog(t *testing.T) {
...
+				close(letTheRaceBegin)
+				SetDestination(io.Discard, nil)
+				<-letTheRaceEnd
+				So(true, ShouldEqual, true) // all we are testing for is the absence of detecting a data race
+			})
+
+		})
+
+		Convey("Given the standard destination returns an error", func() {
+			destination = WriteWillError{}
```
Comment (nshumoogum) [inline log/log_test.go:462]: destination not used, so can be removed

**Review by nshumoogum — CHANGES_REQUESTED** [review]

File: log/log_test.go (hunk 3 above)
Comment (goofballLogic) [reply log/log_test.go:462]: this (and the types above) are used within the log.go file so that writes go to the fallbackDestination instead

**Review by nshumoogum — DISMISSED** [review]

File: log/log_test.go (hunk 3 above)
Comment (nshumoogum) [reply log/log_test.go:462]: Yep misunderstood this bit but looking back over it I can see this is to force any error on the destination so as you say the code falls back onto the "fallbackDestination" and therefore testing the race conditon on this global variable.

**Review by nshumoogum — DISMISSED** [review]

**Review by nshumoogum — APPROVED** [review]

**Pull request merged**